### PR TITLE
Implement builtin debugger intrinsic `__log2`

### PR DESCRIPTION
**Pull request by werat**
Implement builtin function in the parser and the evaluator. Currently only one function is supported -- `__log2`.

* Save source location in the AST node to allow for better error diagnostics
* Implement basic implicit type conversions checking -- this is required for handling function arguments

### CODE REVIEW

File: docs/expr-lang.ebnf [hunk 1]
```diff
@@ -140,3 +141,12 @@ numeric_literal = ? clang::tok::numeric_constant ? ;
 boolean_literal = "true" | "false" ;
 
 pointer_literal = "nullptr" ;
+
+builtin_func = builtin_func_name "(" {builtin_func_argument_list} ")" ;
+
+builtin_func_name = "__log2" ;
+
+builtin_func_argument_list = builtin_func_argument
+                           | builtin_func_argument_list "," builtin_func_argument_list
```
Comment (tonkosi) [inline docs/expr-lang.ebnf:150]: Is this supposed to be `builtin_func_argument "," builtin_func_argument_list`?

File: lldb-eval/parser.cc [hunk 2]
```diff
@@ -1780,7 +1816,102 @@ ExprResult Parser::ParseIntegerLiteral(clang::NumericLiteralParser& literal,
...
+  ConsumeToken();
+
+  Expect(clang::tok::l_paren);
+  ConsumeToken();
+
+  std::vector<ExprResult> arguments;
+
+  do {
+    // Eat the comma if this is not the first iteration.
+    if (arguments.size() > 0) {
```
Comment (jaro-sevcik) [inline lldb-eval/parser.cc:1850]: Nit: I guess you could consume the token unconditionally and remove the ConsumeToken for lparen.

File: docs/expr-lang.ebnf (hunk 1 above)
Comment (jaro-sevcik) [inline docs/expr-lang.ebnf:150]: One of those builtin_func_argument_lists could be builtin_func_argument, no?

Also, is there a reason to forbid empty argument list?

File: lldb-eval/eval_test.cc [hunk 3]
```diff
@@ -1618,3 +1618,41 @@ TEST_F(EvalTest, TestSizeOf) {
...
+  EXPECT_THAT(Eval("__log2(cxx_enum)"), IsEqual("7"));
+  EXPECT_THAT(Eval("__log2(CEnum::kFoo)"), IsEqual("7"));
+  EXPECT_THAT(Eval("__log2(CxxEnum::kFoo)"), IsEqual("7"));
+
+  EXPECT_THAT(Eval("__log2(foo)"),
+              IsError("no known conversion from 'Foo' to 'unsigned int'\n"
+                      "__log2(foo)\n"
+                      "       ^"));
+
+  EXPECT_THAT(Eval("1 + __log2(1, 2)"),
```
Comment (jaro-sevcik) [inline lldb-eval/eval_test.cc:1653]: Also add a test for no arguments, please.

File: lldb-eval/parser.cc [hunk 4]
```diff
@@ -938,6 +966,7 @@ ExprResult Parser::ParseUnaryExpression() {
   }
 
   if (token_.is(clang::tok::kw_sizeof)) {
+    clang::Token token = token_;
```
Comment (tonkosi) [inline lldb-eval/parser.cc:969]: Nit: The source location can be calculated directly here.

File: lldb-eval/eval_test.cc (hunk 3 above)
Comment (jaro-sevcik) [reply lldb-eval/eval_test.cc:1653]: Perhaps also for something >2^32.

File: docs/expr-lang.ebnf (hunk 1 above)
Comment (jaro-sevcik) [reply docs/expr-lang.ebnf:150]: (Looking at the parser, it looks like non-empty arguments was an intentional decision.)

File: lldb-eval/eval_test.cc (hunk 3 above)
Comment (jaro-sevcik) [reply lldb-eval/eval_test.cc:1653]: And maybe the usual double suspects (+/- infinity, nan).

File: lldb-eval/parser.cc [hunk 5]
```diff
@@ -1780,7 +1816,102 @@ ExprResult Parser::ParseIntegerLiteral(clang::NumericLiteralParser& literal,
...
+
+    arguments.push_back(std::move(argument));
+
+  } while (token_.is(clang::tok::comma));
+
+  Expect(clang::tok::r_paren);
+  ConsumeToken();
+
+  // Check we have the correct number of arguments.
+  if (arguments.size() != func_def->arguments_.size()) {
```
Comment (tonkosi) [inline lldb-eval/parser.cc:1869]: We could also bail out a bit earlier in the case there are more arguments than expected with additional check in the do-while loop. Is that going to make a difference?

File: docs/expr-lang.ebnf (hunk 1 above)
Comment (werat) [reply docs/expr-lang.ebnf:150]: Technically the other way around (`builtin_func_argument_list "," builtin_func_argument`), but yes, thanks for catching this!

File: lldb-eval/parser.cc (hunk 4 above)
Comment (werat) [reply lldb-eval/parser.cc]: Done

File: lldb-eval/parser.cc (hunk 2 above)
Comment (werat) [reply lldb-eval/parser.cc]: This is no longer the case when handling empty argument list

File: lldb-eval/parser.cc (hunk 5 above)
Comment (werat) [reply lldb-eval/parser.cc:1877]: We need to parse all the arguments so that we can have an error message like "requires N arguments, got K arguments".

Comment (tonkosi) [reply lldb-eval/parser.cc:1877]: Makes sense, thanks for clarification.

**Review by tonkosi — APPROVED** [review]

**Pull request merged**